src/metadata.rs

```diff
@@ -356,7 +356,13 @@ impl Metadata21 {
             gui_scripts: HashMap::new(),
             entry_points: HashMap::new(),
         };
-        metadata.merge_pyproject_toml(manifest_path)?;
+
+        let manifest_path = manifest_path.as_ref();
+        let py_src = extra_metadata
+            .python_source
+            .map(|src| manifest_path.join(src))
+            .unwrap_or_else(|| manifest_path.to_path_buf());
+        metadata.merge_pyproject_toml(py_src)?;
```

Comment:
To avoid the possible path copy:
```suggestion
        let py_src = extra_metadata.python_source.map_or(
            Cow::Borrowed(manifest_path),
            |src| Cow::Owned(manifest_path.join(src))
        );
        metadata.merge_pyproject_toml(&*py_src)?;
```